File: hummingbird/androidApp/app/src/main/java/net/acumenus/hummingbird/data/TransportSecurityPolicy.kt
package net.acumenus.hummingbird.data

import net.acumenus.hummingbird.BuildConfig
import java.net.URI

enum class StaffTransportEnvironment {
    DEVELOPMENT,
    PRODUCTION;

    companion object {
        fun fromBuild(): StaffTransportEnvironment =
            if (BuildConfig.DEBUG) DEVELOPMENT else PRODUCTION
    }
}

/**
 * Fail-closed transport boundary for the staff product.
 *
 * Release builds have one HTTPS/WSS origin. Development builds may use a
 * system-trusted HTTPS origin or a tightly bounded loopback/emulator cleartext
 * origin. TLS trust remains owned by Android Network Security Configuration and
 * the platform trust store; this policy never installs a custom TrustManager.
 */
object StaffTransportSecurityPolicy {
    const val PRODUCTION_HOST = "zephyrus.acumenus.net"

    private val developmentCleartextHosts = setOf(
        "localhost",
        "127.0.0.1",
        "10.0.2.2",
    )

    fun permitsHttpBaseUrl(
        rawUrl: String,
        environment: StaffTransportEnvironment,
    ): Boolean {
        val uri = runCatching { URI(rawUrl) }.getOrNull() ?: return false
        val scheme = uri.scheme?.lowercase() ?: return false
        val host = uri.host?.lowercase() ?: return false
        if (
            uri.rawUserInfo != null ||
            uri.rawQuery != null ||
            uri.rawFragment != null ||
            (uri.port != -1 && uri.port !in 1..65_535) ||
            !(uri.rawPath.isNullOrEmpty() || uri.rawPath == "/")
        ) {
            return false
        }

        return when (environment) {
            StaffTransportEnvironment.PRODUCTION ->
                scheme == "https" &&
                    host == PRODUCTION_HOST &&
                    (uri.port == -1 || uri.port == 443)
            StaffTransportEnvironment.DEVELOPMENT ->
                scheme == "https" ||
                    (scheme == "http" && host in developmentCleartextHosts)
        }
    }

    fun permitsWebSocket(
        scheme: String,
        host: String,
        port: Int,
        environment: StaffTransportEnvironment,
    ): Boolean {
        if (port !in 1..65_535) return false

        val normalizedScheme = scheme.lowercase()
        val normalizedHost = host.lowercase()

        return when (environment) {
            StaffTransportEnvironment.PRODUCTION ->
                normalizedScheme == "wss" &&
                    normalizedHost == PRODUCTION_HOST &&
                    port == 443
            StaffTransportEnvironment.DEVELOPMENT ->
                (normalizedScheme == "wss" && port == 443) ||
                    (normalizedScheme == "ws" && normalizedHost in developmentCleartextHosts)
        }
    }

    fun requireBuildConfiguration() {
        val environment = StaffTransportEnvironment.fromBuild()
        require(permitsHttpBaseUrl(BuildConfig.ZEPHYRUS_BASE_URL, environment)) {
            "Hummingbird staff API transport configuration is unsafe."
        }
        require(
            permitsWebSocket(
                scheme = BuildConfig.ZEPHYRUS_REVERB_SCHEME,
                host = BuildConfig.ZEPHYRUS_REVERB_HOST,
                port = BuildConfig.ZEPHYRUS_REVERB_PORT,
                environment = environment,
            ),
        ) {
            "Hummingbird staff realtime transport configuration is unsafe."
        }
    }
}

File: hummingbird/iosApp/Hummingbird/Networking/TransportSecurityPolicy.swift
import Foundation

enum HummingbirdTransportEnvironment: Equatable {
    case development
    case production

    static var current: HummingbirdTransportEnvironment {
#if DEBUG
        .development
#else
        .production
#endif
    }
}

/// Fail-closed transport boundary for the staff product.
///
/// Release builds have one HTTPS/WSS origin. Development builds may use a
/// system-trusted HTTPS origin or a tightly bounded loopback/emulator cleartext
/// origin. Certificate validation remains owned by ATS and URLSession; this type
/// never installs a permissive trust delegate.
struct HummingbirdTransportSecurityPolicy {
    static let productionHost = "zephyrus.acumenus.net"

    let environment: HummingbirdTransportEnvironment

    func permitsHTTPBaseURL(_ url: URL) -> Bool {
        guard let components = normalizedComponents(url),
              let scheme = components.scheme?.lowercased(),
              let host = components.host?.lowercased()
        else { return false }
        if let port = components.port, !(1...65_535).contains(port) {
            return false
        }

        switch environment {
        case .production:
            return scheme == "https"
                && host == Self.productionHost
                && (components.port == nil || components.port == 443)
        case .development:
            if scheme == "https" {
                return true
            }
            return scheme == "http"
                && Self.developmentCleartextHosts.contains(host)
        }
    }

    func permitsWebSocket(scheme: String, host: String, port: Int) -> Bool {
        guard (1...65_535).contains(port) else { return false }

        let normalizedScheme = scheme.lowercased()
        let normalizedHost = host.lowercased()

        switch environment {
        case .production:
            return normalizedScheme == "wss"
                && normalizedHost == Self.productionHost
                && port == 443
        case .development:
            if normalizedScheme == "wss" {
                return port == 443
            }
            return normalizedScheme == "ws"
                && Self.developmentCleartextHosts.contains(normalizedHost)
        }
    }

    private func normalizedComponents(_ url: URL) -> URLComponents? {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              components.user == nil,
              components.password == nil,
              components.query == nil,
              components.fragment == nil,
              components.path.isEmpty || components.path == "/"
        else { return nil }

        return components
    }

    private static let developmentCleartextHosts: Set<String> = [
        "localhost",
        "127.0.0.1",
        "::1",
        "10.0.2.2",
    ]
}

/// Refuses every HTTP redirect so an otherwise valid API or realtime origin
/// cannot move a credential-bearing request to a second endpoint.
final class HummingbirdNoRedirectDelegate: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

enum HummingbirdURLSessionFactory {
    static func make(
        configuration: URLSessionConfiguration = .ephemeral
    ) -> URLSession {
        URLSession(
            configuration: configuration,
            delegate: HummingbirdNoRedirectDelegate(),
            delegateQueue: nil
        )
    }
}
